### PR TITLE
fix(date): ensure correct output format and input formats are used for given locale FE-5035

### LINKED ISSUE #4966: ISO date format not supported
### Current behaviour

When setting a **zh-CN** localization, the date input reformats to the day-month-year format instead of ISO 8601.

### Expected behaviour

When setting a **zh-CN** localization, the date input reformats to the right format: ISO 8601.

### CodeSandbox or Storybook URL

https://codesandbox.io/s/ancient-wave-0d3m3t

### JIRA Ticket (Sage Only)

XT-21014

### Suggested Solution

I think new hardcoded language tags (at least `zh-CN`, but maybe `pl`) can be added to the *src/components/date/_\_internal__/date-formats/index.js* `getFormatDate` function. Also, some ISO input formats as you have for EU_FORMATS and NA_FORMATS.

### Carbon Version

104.27.0

### What browsers are you seeing the problem on?

Chrome

### What Operating System are you seeing the problem on?

Windows

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### LINKED ISSUE #5072: DateInput not raising OnBlur event
### Current behaviour

DateInput does not reliably raise the OnBlur event.

eg:
Click outside date control. 
Click inside date control. (all characters highlight blue)
Click in date control at end of string, to show caret, and with characters no longer highlight blue
Use backspace to delete each character individually - delete all of them
Tab outside date control.
onChange raised as each char deleted.
But onBlur NOT raised on tabbing outside the control.

Sandbox attached.

### Expected behaviour

OnBlur to be raised on leaving the control: so that we can update our state (yyyy-mm-dd) and the visible (dd/mm/yyyy) versions of the date, so that they are consistent. 

### CodeSandbox or Storybook URL

https://codesandbox.io/s/cool-danny-ipsc0j

### JIRA Ticket (Sage Only)

_No response_

### Suggested Solution

_No response_

### Carbon Version

104.33

### What browsers are you seeing the problem on?

Chrome

### What Operating System are you seeing the problem on?

Windows

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### LINKED ISSUE #5119: The rawValue emitted by Date components is incorrect when year string is only 2 digits
### Current behaviour

When the the year part of the input string is only 2 digits the event values emitted do not match the input's value: the rawValue is also different from formattedValue onBlur

Can be seen if you input `12.12.69` into linked sandbox

### Expected behaviour

Values emitted should match

### CodeSandbox or Storybook URL

https://codesandbox.io/s/quizzical-sky-i0zhq2?file=/src/index.js

### JIRA Ticket (Sage Only)

_No response_

### Suggested Solution

_No response_

### Carbon Version

latest

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Microsoft Edge

### What Operating System are you seeing the problem on?

MacOS

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### PR DESCRIPTION
**Pull request by edleeks87**
fix #4966
fix #5072
fix #5119

### Proposed behaviour

Ensures that the correct formats are used for input and output strings. Fixes the output string and
valid inputs for Chinese (CN) locales for example by using the [Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) util API to handle the format displayed
on blur of the text input

image

Ensures that the the callback passed to the `onBlur` prop is called when the text input is blurred.

Ensure that the rawValue matches formattedValue when a year string is only 2 digits long
image

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] Unit tests added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

Two issue have sandbox repros that can be tested

QA should also test DateRange against current master to ensure the blur behaviour remains consistent

https://codesandbox.io/s/carbon-quickstart-forked-vwnj65?file=/src/index.js

### CODE REVIEW

File: src/components/date/__internal__/date-formats/index.js [hunk 1]
```diff
@@ -34,8 +34,31 @@ const NA_FORMATS = [
   "MM dd yyyy",
 ];
 
+// The order of this array is important
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/date/__internal__/date-formats/index.js:37]: Maybe that's a good suggestion 😉

File: src/components/date/__internal__/date-formats/index.js [hunk 2]
```diff
@@ -47,24 +70,61 @@ const generateFormats = (formatArray) =>
...
+    .formatToParts(new Date())
+    .map(({ type, value }) => {
+      if (type !== "literal") return formatMap[type];
+
+      return value;
+    })
+    .join("");
+
+  if (localeCode === "bg-BG") {
+    // this locale adds an additional char that has no effect on the output formatting
```
Comment (Parsium) [inline src/components/date/__internal__/date-formats/index.js:92]: Are there any other locales we support that add additional characters that we would need to cover?

**Review by Parsium — DISMISSED** [review]

**Review by Parsium — COMMENTED** [review]
LGTM, happy to approve once @DipperTheDan's comment has been resolved 👍🏼

**Review by DipperTheDan — APPROVED** [review]

**Review by Parsium — APPROVED** [review]

**Review by DipperTheDan — APPROVED** [review]

File: src/components/date/__internal__/date-formats/index.js [hunk 3]
```diff
@@ -34,37 +34,122 @@ const NA_FORMATS = [
...
+  const formatter = new Intl.DateTimeFormat(localeCode);
+  let separator;
+
+  const format = formatter
+    .formatToParts(new Date())
+    .map(({ type, value }) => {
+      if (type !== "literal") {
+        return formatMap[type];
+      }
+      if (!separator) {
```
Comment (Parsium) [inline src/components/date/__internal__/date-formats/index.js:97]:
```suggestion
      else if (!separator) {
```

Not sure if this needs to be an `else` so `separator` is only set to the `value` corresponding to the first `type: literal` found in the array, so its guaranteed `separator` will be set to a separator character and not one associated with a day, month, year, etc.?

Comment (edleeks87) [reply src/components/date/__internal__/date-formats/index.js:97]: The `return` statement on line 95 would prevent the `if` being reached so there's no need for the `else`

**Review by Parsium — APPROVED** [review]

**Review by DipperTheDan — DISMISSED** [review]

**Review by DlgSHi — DISMISSED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by Parsium — APPROVED** [review]

**Pull request merged**